### PR TITLE
Fix pattern for tests that use HttpClient

### PR DESCRIPTION
**Pull request by rynowak**
# Description

This change introduces a better API for us to test the DaprClient or
other HttpClient based APIs.

This will resolve the flakiness problems that we're seeing with some of
the actors tests.

Additionally fixed an issue where DaprHttpInteractor was misuing
HttpClientHandler. This would result in a new handler being created when
it isn't needed.

## Issue reference

Fixes: #573
Fixes: #588

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

File: src/Dapr.Actors/Client/ActorProxyFactory.cs [hunk 1]
```diff
@@ -32,7 +32,16 @@ public ActorProxyOptions DefaultOptions
...
+        [Obsolete("Use the constructor that accepts HttpMessageHandler. This will be removed in the future.")]
+        public ActorProxyFactory(ActorProxyOptions options, HttpClientHandler handler)
+            : this(options, (HttpMessageHandler)handler)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ActorProxyFactory"/> class.
+        /// </summary>
+        public ActorProxyFactory(ActorProxyOptions options = null, HttpMessageHandler handler = null)
```
Comment (rynowak) [inline src/Dapr.Actors/Client/ActorProxyFactory.cs:44]: The way that this API was design isn't very flexible/testable. `HttpClientHandler` isn't a more specific type than `HttpMessageHandler`. 

The API change in this PR is necessary to make this code testable, and is done in a way that's binary-compatible. It's safe to remove a default value for an optional parameter as long as you provide a *better* overload. Only users that were explicitly passing a `HttpClientHandler` will be impacted by this, and they could just choose to suppress the warning.

File: src/Dapr.Actors/DaprHttpInteractor.cs [hunk 2]
```diff
@@ -28,19 +28,20 @@ internal class DaprHttpInteractor : IDaprInteractor
         private readonly JsonSerializerOptions jsonSerializerOptions = JsonSerializerDefaults.Web;
         private const string DaprEndpoint = Constants.DaprDefaultEndpoint;
         private readonly string daprPort;
-        private static HttpClientHandler innerHandler = new HttpClientHandler();
-        private HttpClient httpClient = null;
-        private bool disposed = false;
+        private readonly static HttpMessageHandler defaultHandler = new HttpClientHandler();
```
Comment (rynowak) [inline src/Dapr.Actors/DaprHttpInteractor.cs:31]: This was the root cause of #588 and possibly also of #573.

Comment (rynowak) [reply src/Dapr.Actors/DaprHttpInteractor.cs:31]: ⚠️ ⚠️ ⚠️ mutable statics and dangerous ⚠️ ⚠️ ⚠️

File: test/Dapr.Actors.Test/ApiTokenTests.cs [hunk 3]
```diff
@@ -16,107 +14,96 @@ namespace Dapr.Actors.Test
...
-            handler.Requests.TryDequeue(out var entry).Should().BeTrue();
-            Action action = () => entry.Request.Headers.GetValues("dapr-api-token");
-            action.Should().Throw<InvalidOperationException>();
+
+            var request = await client.CaptureHttpRequestAsync(async handler =>
+            {
+                var factory = new ActorProxyFactory(null, handler);
+                var proxy = factory.CreateActorProxy<ITestActor>(actorId, "TestActor");
+                await proxy.SetCountAsync(1, new CancellationToken());
+            });
```
Comment (rynowak) [inline test/Dapr.Actors.Test/ApiTokenTests.cs:53]: This design approach lets us capture HTTP or gRPC traffic without timing issues. This is very robust and applies timeouts to everything that happens, so we can't end up with tests that hang indefinitely.

File: test/Dapr.Actors.Test/ApiTokenTests.cs [hunk 4]
```diff
@@ -16,107 +14,96 @@ namespace Dapr.Actors.Test
...
+
+            request.Dismiss();
```
Comment (rynowak) [inline test/Dapr.Actors.Test/ApiTokenTests.cs:55]: The other cool trick here is that we'll throw if you fail to *acknowledge* a request. This way we make sure that extra stuff isn't happening that's going untested.

**Comment by rynowak** [comment]
@vinayada1 - suggestion here is that we hold this for a 1.0.1 patch of the SDK and release it as soon as the dust settles from 1.0.0.

**Review by vinayada1 — APPROVED** [review]

**Pull request merged**